src/commands/validate.ts

```diff
@@ -29,6 +33,22 @@ export default class Validate extends Command {
       specWatcher({ spec: specFile, handler: this, handlerName: 'validate' });
     }
 
-    await validate(this, specFile, flags);
+    const result = await validate(this, specFile, flags);
+
+    try {
+      // Metrics recording.
+      const {document} = await this.parser.parse(specFile.text());
+      if (document !== undefined) {
+        const metadata = MetadataFromDocument(document);
+        metadata['success'] = true;
+        metadata['validation_result'] = result;
+        await this.recorder.recordActionExecution('validate', metadata);
+        await this.recorder.flush();
+      }
+    } catch (e: any) {
+      if (e instanceof Error) {
+        this.log(`Skipping submitting anonymous metrics due to the following error: ${e.name}: ${e.message}`);
+      }
+    }
```

Comment:
Can't we just move this logic to `../parser`? Otherwise, we'd be parsing twice, right?

Reply:
In this case (`validate` cmd), there is no call to `../parser.ts:parse()` method (which internally calls `parser.parse()`) but just `../parser.ts:parse()`, which it doesn't really parse the document into a known model but just validates the document by calling `parser.validate()`.
In short, we need to transform the doc into a known model (Parser-API v1, v2, etc) so we can later extract the data we need for the metrics.

There are other commands that, at some point, call the  `../parser.ts:parse()` methd, such as the `generate models` or `diff`. We could add a method there that calls the parser without logging errors, even though it's not a big improvement in terms of reusing code.
Those methods are logging validation errors to the stderr of the cmd, which is not a desirable behaviour here since we just want to log metrics in an invisible way.

I hope I answered your question/concern.

Reply:
That makes sense. Thanks!